BUG: iterating on a subset of columns in a GroupBy object (#44821)

maurosilber opened the pull request:
Fixes issue #44821.

When trying to iterate on a subset of columns in a GroupBy object,
it returned all columns, instead of the selected subset.

GroupBy.__iter__ used self.obj instead of self._selected_obj (see
PR #6570).

- [x] closes #44821
- [x] 1 test added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

maurosilber committed:
BUG: iterating on a subset of columns in a GroupBy object (#44821)

Fixes issue #44821.

When trying to iterate on a subset of columns in a GroupBy object,
it returned all columns, instead of the selected subset.

GroupBy.__iter__ used self.obj instead of self._selected_obj (see
PR #6570).

phofl reviewed (changes requested):
Please run the tests locally first. This broke lots of tests

maurosilber commented:
Hi.

Only one test is broken, which I mentioned it here: https://github.com/pandas-dev/pandas/issues/44821#issuecomment-997037279

As mentioned there, I'm not sure how to continue, as users may rely on current behavior.

By the way, should I could continue the discussion in this PR?

maurosilber committed:
BUG: fix test in rolling groupby (#44821)

Fixes test due to changes in GroupBy.__iter__ (see #44821).

As the column `c` wasn't selected on the manual computation,
it failed when trying to set it as an index.

maurosilber commented:
Well, I decided to fix that test. While this PR changes `GroupBy.__iter__`, it didn't actually broke any test that tested its behavior. The test that was previously failing concerned the behaviour of `GroupBy.rolling`, and the fix was in the "manual computation" to compare against.

So, although this PR might break somebody's code, this `GroupBy` behavior wasn't "promised" anywhere, and it could have been changed as part of #6570.

jreback reviewed (changes requested):
ok this is reasonable, can you add a whatsnew note in 1.4, groupby bug fix section. you may want to give a mini-example of what the change is

jreback commented:
cc @rhshadrach for comments

maurosilber committed:
Updated whatsnew 1.4.0

rhshadrach commented on doc/source/whatsnew/v1.4.0.rst:
```diff
@@ -797,6 +797,7 @@ Groupby/resample/rolling
 - Bug in :meth:`GroupBy.nth` failing on ``axis=1`` (:issue:`43926`)
 - Fixed bug in :meth:`Series.rolling` and :meth:`DataFrame.rolling` not respecting right bound on centered datetime-like windows, if the index contain duplicates (:issue:`3944`)
 - Bug in :meth:`Series.rolling` and :meth:`DataFrame.rolling` when using a :class:`pandas.api.indexers.BaseIndexer` subclass that returned unequal start and end arrays would segfault instead of raising a ``ValueError`` (:issue:`44470`)
+- Fixed bug in :meth:`GroupBy.__iter__` after selecting a subset of columns in a :class:`GroupBy` object, which returned all columns instead of the chosen subset (:issue:`#44821`).
```
nit: No period after the issue number.

maurosilber replied:
Done.

rhshadrach reviewed (changes requested):
Small nitpick, otherwise I think this is good.

There is an oddity with _selected_obj, but I think this change is definitely an improvement over current behavior and it is really independent, so still +1. Once you subset the columns of a groupby (e.g. `['b']` or `[['b']]`), the selected_obj will not include the grouper. However, before you do this, it will unless the context is active. I.e.

```
df = pd.DataFrame({'a': [1, 2], 'b': [3, 4], 'c': [5, 6]})
gb = df.groupby('a')
print(list(gb.grouper.get_iterator(gb._selected_obj, axis=gb.axis)))
print()
with gb._group_selection_context():
    print(list(gb.grouper.get_iterator(gb._selected_obj, axis=gb.axis)))
    print()
selected_gb = gb[['b']]
print(list(selected_gb.grouper.get_iterator(selected_gb._selected_obj, axis=gb.axis)))
```

produces

```
[(1,    a  b  c
0  1  3  5), (2,    a  b  c
1  2  4  6)]

[(1,    b  c
0  3  5), (2,    b  c
1  4  6)]

[(1,    b
0  3), (2,    b
1  4)]
```

In the first print, the grouper 'a' is present. In the 2nd and 3rd it isn't.

maurosilber committed:
Merge branch 'master' into master

rhshadrach reviewed (approved):
lgtm

jreback commented:
thanks @maurosilber
